pandas/core/ops.py

```diff
@@ -130,6 +130,13 @@ def maybe_upcast_for_op(obj):
         # implementation; otherwise operation against numeric-dtype
         # raises TypeError
         return pd.Timedelta(obj)
+    elif isinstance(obj, np.timedelta64) and not isna(obj):
+        # In particular non-nanosecond timedelta64 needs to be cast to
+        #  nanoseconds, or else we get undesired behavior like
+        #  np.timedelta64(3, 'D') / 2 == np.timedelta64(1, 'D')
+        # The isna check is to avoid casting timedelta64("NaT"), which would
+        #  return NaT and incorrectly be treated as a datetime-NaT.
```

Comment:
Yes.  In the past when I've brought up the idea of having something like `pd.NaTD` the idea has gone nowhere.  I'd be on board if you want to revive the suggestion.

Reply:
Was testing a few things on master, this seems a bug:

```
In [9]: pd.Series([pd.Timedelta('1s')]) + np.timedelta64('nat')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-f56b47937d6f> in <module>()
----> 1 pd.Series([pd.Timedelta('1s')]) + np.timedelta64('nat')

~/scipy/pandas/pandas/core/ops.py in wrapper(left, right)
   1408 
   1409         elif is_timedelta64_dtype(left):
-> 1410             result = dispatch_to_index_op(op, left, right, pd.TimedeltaIndex)
   1411             return construct_result(left, result,
   1412                                     index=left.index, name=res_name,

~/scipy/pandas/pandas/core/ops.py in dispatch_to_index_op(op, left, right, index_class)
   1040         left_idx = left_idx._shallow_copy(freq=None)
   1041     try:
-> 1042         result = op(left_idx, right)
   1043     except NullFrequencyError:
   1044         # DatetimeIndex and TimedeltaIndex with freq == None raise ValueError

~/scipy/pandas/pandas/core/indexes/datetimelike.py in __add__(self, other)
    578         def __add__(self, other):
    579             # dispatch to ExtensionArray implementation
--> 580             result = super(cls, self).__add__(other)
    581             return wrap_arithmetic_op(self, other, result)
    582 

~/scipy/pandas/pandas/core/arrays/datetimelike.py in __add__(self, other)
    626                 result = self._add_nat()
    627             elif isinstance(other, (Tick, timedelta, np.timedelta64)):
--> 628                 result = self._add_delta(other)
    629             elif isinstance(other, DateOffset):
    630                 # specifically _not_ a Tick

~/scipy/pandas/pandas/core/arrays/timedeltas.py in _add_delta(self, delta)
    204         result : TimedeltaArray
    205         """
--> 206         new_values = dtl.DatetimeLikeArrayMixin._add_delta(self, delta)
    207         return type(self)(new_values, freq='infer')
    208 

~/scipy/pandas/pandas/core/arrays/datetimelike.py in _add_delta(self, other)
    371         """
    372         if isinstance(other, (Tick, timedelta, np.timedelta64)):
--> 373             new_values = self._add_timedeltalike_scalar(other)
    374         elif is_timedelta64_dtype(other):
    375             # ndarray[timedelta64] or TimedeltaArray/index

~/scipy/pandas/pandas/core/arrays/datetimelike.py in _add_timedeltalike_scalar(self, other)
    383         return the i8 result view
    384         """
--> 385         inc = delta_to_nanoseconds(other)
    386         new_values = checked_add_with_arr(self.asi8, inc,
    387                                           arr_mask=self._isnan).view('i8')

~/scipy/pandas/pandas/_libs/tslibs/timedeltas.pyx in pandas._libs.tslibs.timedeltas.delta_to_nanoseconds()

~/scipy/pandas/pandas/_libs/tslibs/timedeltas.pyx in pandas._libs.tslibs.timedeltas.delta_to_nanoseconds()

TypeError: an integer is required
```

But if I remove here the `not isna` check, it works correctly (since `s + pd.NaT` works correctly and preserves the timedelta dtype). But I suppose there are other code paths don't work with the `pd.NaT` ?

Reply:
Good catch.  I'll get a PR to fix+test this going sometime today.